tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11238,1067 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+  // 4. Optional: -DFUNC : min, max, etc.
+  std::string ShaderText = R"(
+  struct LongVectorOp{
+    TYPE ScalarInput;
+    vector<TYPE, NUM> VecInput1;
+    vector<TYPE, NUM> VecInput2;
+    vector<TYPE, NUM> VecOutput;
+  };
+
+  vector<TYPE, NUM> testInitialize(vector<TYPE, NUM> Vec)
+  {
+    vector<TYPE, NUM> VecCopy = Vec;
+    return VecCopy;
+  }
+
+  vector<TYPE, NUM> testClamp(vector<TYPE, NUM> Vec)
+  {
+    return clamp(Vec, TYPE(CLAMP_ARGC), TYPE(CLAMP_ARGT));
+  }
+
+  RWStructuredBuffer<LongVectorOp> g_buf : register(u0);
+  [numthreads(8,8,1)]
+  void main(uint GI : SV_GroupIndex) {
+    LongVectorOp LVOp = g_buf[GI];
+    LVOp.VecOutput = FUNC(LVOp.VecInput1 OPERATOR OPERAND2);
+    g_buf[GI] = LVOp;
+  };
+  )";
+
+  // Set up the compiler options string.
+  std::stringstream CompilerOptions("");
+  std::string HLSLType = TestConfig.GetHLSLTypeString();
+  CompilerOptions << "-DTYPE=";
+  CompilerOptions << HLSLType;
+  CompilerOptions << " -DNUM=";
+  CompilerOptions << N;
+  const bool Is16BitType = (HLSLType == "int16_t" || HLSLType == "uint16_t" || HLSLType == "half");
+  CompilerOptions << (Is16BitType ? " -enable-16bit-types" : "");
+  CompilerOptions << " -DOPERATOR=";
+  CompilerOptions << TestConfig.OperatorString;
+  CompilerOptions << " -DOPERAND2=";
+  if (TestConfig.IsBinaryOp) {
+    CompilerOptions<< (TestConfig.IsScalarOp ? "LVOp.ScalarInput" : "LVOp.VecInput2");
+  }
+  CompilerOptions << " -DFUNC=";
+  CompilerOptions << TestConfig.IntrinsicString;
+  CompilerOptions << " -DCLAMP_ARGC=";
+  CompilerOptions << ClampArgC;
+  CompilerOptions << " -DCLAMP_ARGT=";
+  CompilerOptions << ClampArgT;
+
+  // We have to construct the string outside of the lambda. Otherwise it's
+  // cleaned up when the lambda finishes executing but before the shader runs.
+  std::string CompilerOptionsString = CompilerOptions.str();
+
+  // ShaderOpArith.xml defines some boiler plate for the shader code.
+  CComPtr<IStream> TestXML;
+  ReadHlslDataIntoNewStream(L"ShaderOpArith.xml", &TestXML);
+
+  // RunShaderOpTest is a helper function that does all device/resource creation
+  // and setup. It also handles the shader compilation and execution. It takes a
+  // callback that is called when the shader is compiled, but before it is
+  // executed.
+  std::shared_ptr<ShaderOpTestResult> TestResult = RunShaderOpTest(
+    D3DDevice, m_support, TestXML, "LongVectorOp", 
+    [&](LPCSTR Name, std::vector<BYTE> &ShaderData, st::ShaderOp *ShaderOp) {
+      LogCommentFmt(L"RunShaderOpTest CallBack. Shader Name: %S", Name);
+
+      // Sanity. Helps diagnose test test configuration issues.
+      VERIFY_IS_TRUE(0 == _stricmp(Name, "LongVectorOp"));
+
+      const size_t DataSize = sizeof(LongVectorOp<T, N>);
+      ShaderData.resize(DataSize);
+
+      // Treat the data as an array of LongVectorOp<T, N> structs.
+      LongVectorOp<T, N> *ShaderStruct = reinterpret_cast<LongVectorOp<T, N> *>(ShaderData.data());
```

Comment:
This casting logic isn't robust for a struct, nor is the copy logic below robust when copying HLSLBool_t or HLSLHalf_t. It's working fine for me for now. But I intend to swap to ByteAddressBuffers for input/output. I'll make this more robust when I do that. I'll do that for this PR.

Reply:
Fixed in latest revision.